src/System.Text.Json/tests/Serialization/ExtensionDataTests.cs

```diff
@@ -178,23 +178,54 @@ public static void IgnoredDataShouldNotBeExtensionData()
         }
 
         [Fact]
-        public static void InvalidExtensionValue()
+        public static void ExtensionPropertyObjectValue_Empty()
+        {
+            // Baseline
+            ClassWithExtensionPropertyAlreadyInstantiated obj = JsonSerializer.Parse<ClassWithExtensionPropertyAlreadyInstantiated>(@"{}");
+            Assert.Equal(@"{""MyOverflow"":{}}", JsonSerializer.ToString(obj));
+        }
+
+        [Fact]
+        public static void ExtensionPropertyObjectValue()
         {
             // Baseline
             ClassWithExtensionPropertyAlreadyInstantiated obj = JsonSerializer.Parse<ClassWithExtensionPropertyAlreadyInstantiated>(@"{}");
             obj.MyOverflow.Add("test", new object());
+            obj.MyOverflow.Add("test1", 1);
 
-            try
-            {
-                JsonSerializer.ToString(obj);
-                Assert.True(false, "InvalidOperationException should have thrown.");
-            }
-            catch (InvalidOperationException e)
-            {
-                // Verify the exception contains the property name and invalid type.
-                Assert.Contains("ClassWithExtensionPropertyAlreadyInstantiated.MyOverflow", e.Message);
-                Assert.Contains("System.Object", e.Message);
-            }
+            Assert.Equal(@"{""MyOverflow"":{""test"":{},""test1"":1}}", JsonSerializer.ToString(obj));
+        }
+
+        [Fact]
+        public static void ExtensionPropertyObjectValue_RoundTrip()
+        {
+            // Baseline
+            ClassWithExtensionPropertyAlreadyInstantiated obj = JsonSerializer.Parse<ClassWithExtensionPropertyAlreadyInstantiated>(@"{}");
+            obj.MyOverflow.Add("test", new object());
+            obj.MyOverflow.Add("test1", 1);
+            obj.MyOverflow.Add("test2", "text");
+            obj.MyOverflow.Add("test3", new DummyObj() { Prop = "Prop" });
```

Comment:
So, `JsonExtensionData` on `Dictionary<string, object>` and `Dictionary<string, JsonElement>` is supported, but nothing else.

nit: When it is `Dictionary<string, object>`, the values can be any .NET object. Here, you have added object, int, string, and a custom class. Can you also add a collection (like another Dictionary), and also another struct?

Reply:
> So, JsonExtensionData on Dictionary<string, object> and Dictionary<string, JsonElement> is supported, but nothing else.

Yes https://github.com/dotnet/corefx/blob/a78bd7308555926f88aad3d0ecaaebae0f9204cc/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleDictionary.cs#L38

Reply:
I'll add tests.

Reply:
done